### PR TITLE
feat(api): Delete redundant class parameter from `Config.get()` when using with default config

### PR DESCRIPTION
**Pull request by cb0s**
### CLA

- [x] I have signed the individual contributor's license agreement **and sent** it to the board of the WüSpace e. V. organization.

### CODE REVIEW

**Comment by fussel178** [comment]
Well, I think this is a breaking change.
And when we can determine the class type via the default configuration, we can also determine the class type via the forced configuration. Can you also change the API for the `on` method without default configuration?

**Review by fussel178 — CHANGES_REQUESTED** [review]
Well, I think this is a breaking change.
And when we can determine the class type via the default configuration, we can also determine the class type via the forced configuration.
Can you also change the API for the `on` method without default configuration to use `forcedConfig.getClass()` instead of `Class<T> type`?

**Comment by cb0s** [comment]
> 
> 
> And when we can determine the class type via the default configuration, we can also determine the class type via the forced configuration. Can you also change the API for the `on` method without default configuration to use `forcedConfig.getClass()` instead of `Class<T> type`?

I am afraid this is not possible, because we can set forcedConfig to `null` which is the case, when there is not default config.

When you use a default config however, you usually don't provide a null argument (which does not have a `getClass()` method) because then you could also use the "normal" `get()` without a default config.

File: modules/telestion-api/src/main/java/de/wuespace/telestion/api/config/Config.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package de.wuespace.telestion.api.config;
 
+import io.reactivex.annotations.NonNull;
```
Comment (jvpichowski) [inline modules/telestion-api/src/main/java/de/wuespace/telestion/api/config/Config.java:3]: Huhh? Reactivex???

File: modules/telestion-api/build.gradle [hunk 2]
```diff
@@ -52,6 +52,7 @@ dependencies {
     implementation group: 'net.logstash.logback', name: 'logstash-logback-encoder', version: '6.6'
 
     implementation "io.vertx:vertx-core:4.2.1"
+    implementation "io.vertx:vertx-rx-java2:4.2.1"
```
Comment (jvpichowski) [inline modules/telestion-api/build.gradle:55]: Hugh?? Rx-Java?

File: modules/telestion-api/src/main/java/de/wuespace/telestion/api/config/Config.java [hunk 3]
```diff
@@ -27,14 +28,16 @@
...
+	 *                      it must not be <code>null</code>!
 	 * @param jsonConfig    this is common config which represents the config class in json. If fields are not available
 	 *                      they will be set like the default values of the class when the default constructor is called.
-	 * @param type          the type of the config.
 	 * @param <T>           defines the class type of the configuration
 	 * @return the selected configuration
 	 */
-	public static <T> T get(T forcedConfig, T defaultConfig, JsonObject jsonConfig, Class<T> type) {
+	@SuppressWarnings("unchecked")
+	public static <T> T get(T forcedConfig, @NonNull T defaultConfig, JsonObject jsonConfig) {
```
Comment (jvpichowski) [inline modules/telestion-api/src/main/java/de/wuespace/telestion/api/config/Config.java:39]: I guess it brakes all other code.

Why should we rely on unsafe code and revert a common pattern? 

You need to explain, how this impacts our code base in a good way. At the moment I don't see it.

**Review by jvpichowski — DISMISSED** [review]
I don't see the benefit and there are some strange imports.

**Comment by jvpichowski** [comment]
> This is a "special case" where it is unnecessary and that's why I thought, we could delete it. 

In no way. But maybe you could introduce a new method named "unsafe", if you really love to do it this way. But before, show me a use case. Usually I want to see three use cases before I start changing the core. So I'm interested what you have :)

File: modules/telestion-api/src/main/java/de/wuespace/telestion/api/config/Config.java (hunk 1 above)
Comment (cb0s) [reply modules/telestion-api/src/main/java/de/wuespace/telestion/api/config/Config.java:3]: This belongs to vertx :)

File: modules/telestion-api/src/main/java/de/wuespace/telestion/api/config/Config.java (hunk 3 above)
Comment (cb0s) [reply modules/telestion-api/src/main/java/de/wuespace/telestion/api/config/Config.java:39]: Yes it brakes stuff... However this allows for simpler code as `Class<T> type` is redundant for a `defaultConfig` which is not `null`.

As we usually call `Config.get(this.config, new Configuration(), config(), Configuration.class)` we could shorten this to `Config.get(this.config, new Configuration(), config())` as the defaultConfig is never `null` in that case.

Would you rather see an `unsafeGet()`?

**Comment by pklaschka** [comment]
I don't have strong feelings about this but will say that I prefer it the way it is now (since it's easier to differentiate when reading the code).

If we want to change this, I would like to propose to wait until we have to have breaking changes, anyways (that way, we can keep `main` releasable without breaking changes, for now).

**Comment by cb0s** [comment]
I have implemented a getUnsafe which should be called if that what is done is working for sure... This allows for slimmer code. For the legacy structure: instead of calling 
```
this.config = Config.get(this.forcedConfig, new Configuration(), this.config(), Configuration.class);
```
one could call
```
this.config = Config.getUnsafe(this.forcedConfig, new Configuration(), this.config());
```
which reduces the static method on the "important stuff" - the contents of the Configuration.

Do you want to implement this change or close the PR as the new structure is already implemented in the main branch?

**Comment by fussel178** [comment]
We should close this PR because the new `TelestionVerticle` is already merged and the `Config` interface is now deprecated.
But as always, thank you for your invested time and work! :wink:

**Comment by cb0s** [comment]
Ok, then let's work on the new features :D
